### ISSUE TITLE
Resources located in linked framework can't be referenced directly

### ISSUE THREAD

**Issue by RomainBoulay** [tuist/tuist#1779]
**Describe the bug**

Tuist throws an error when a resource located in a linked framework is referenced from another target.

**Steps to reproduce**

The overall architecture of the project:

```
(AppProject)Target1 ---links to---> FrameworkA.framework
```

FrameworkA file structure is something like:

- FrameworkA.framework
  - File1.js
  - FrameworkA (binary)
  - Resource1.png
  - ...

And the app that links to this framework via this line:

```swift
// In AppProject's `Project.swift` file:

Target(name: "Target1",
		...,
		dependencies: [(...), TargetDependency.framework(path: <path_to_frameworkA>)-,
		...)
```

Everything works as expected until then. But, if `Target1` needs to know about a specific resource located in `FrameworkA`, like this:

```swift
// In ForeignProject's `Project.swift` file:

Target(name: "Target1",
		...,
		resources: [<path to app resource1>, <path to app resource2>, <path_to_frameworkA>/File1.js]
		dependencies: [(...), TargetDependency.framework(path: <path_to_frameworkA>)-,
		...)
```

... then `tuist` errors when at generation time (i.e. `tuist generate/focus`)

**Expected behavior**

Tuist does not error when a resource located in a linked framework is referenced from another target.

**Desktop (please complete the following information):**

- OS: macOS
- Version 1.15.0

**Comment by kwridan** [comment]
Thanks for flagging this @RomainBoulay - this seems to be a recurring issue :/

- https://github.com/tuist/tuist/issues/1584
- https://github.com/tuist/tuist/issues/1334#issuecomment-626901732

The previous use cases we've seen being reported were for bundles offered by third party static frameworks via Carthage. Curious as to the use case you are encountering? I would have thought framework internals would be treated as private to that framework. Typically framework resources would be loaded from the framework using `Bundle(for: FrameworkA.SomeClass.self)`.

You can work around this using symlinks, but worth re-evaluating if this is needed in the first place :)

**Comment by RomainBoulay** [comment]
Thanks @kwridan for your quick answer!!

Indeed the dependency is a static Carthage framework. From what I can remember, `Bundle(for: FrameworkA.SomeClass.self)` didn't work but I'll double-check and post more details!

**Comment by RomainBoulay** [comment]
So I just had a look and the code crashes when a class inside `FrameworkA` tries to access one of its resources... 

```swift
Bundle(for: ClassInFrameworkA.self).path(forResource: "File1", ofType: "js")
```

The way we are working around that is by copying this resource `File1.js` inside `Target1`... and I'm very surprised it actually works.

Since we are forcing Carthage dependencies to be compiled as static frameworks, the issue might be caused by that, but I'm not sure exactly why. Would you have a better understanding of what's causing this problem @kwridan?

**Comment by kwridan** [comment]
> Since we are forcing Carthage dependencies to be compiled as static frameworks, the issue might be caused by that

Ah that explains it, yes when forcing static frameworks, the frameworks no longer support hosting resources. As such, manually including its resources works around that issue. 

As things stand this is a bit tricky to support in Tuist within a single project, we could re-evaluate this if this is a really common use case. 

Some things I could suggest:

The symlink workaround `ln -s MyPrecompiledFramework.framework MyPrecompiledFrameworkResouces` and point the `resources:[]` to the symlink, that way the paths don't clash. This is probably the simplest.

Another is to create a dedicated bundle targets for each of the precompiled frameworks in a separate project _(that way paths won't clash either)_:

```swift
Target(
   name: "MyPrecompiledFrameworkResources",
   product: .bundle,
   resources: [
      "MyPrecompiledFramework.framework/*.js"
   ]
)
```

And then adding `.project(target: "MyPrecompiledFrameworkResouces", path: "//Frameworks")` as a dependency - this will require changing the the `accessor` to leverage the mechanism needed to load bundles (which is a bit longer than usual).

```swift
  class BundleFinder { }
  func precompliedFrameworkBundle() -> Bundle {
        let hostBundle = Bundle(for: BundleFinder.self)
        let path = hostBundle.path(forResource: "MyPrecompiledFrameworkResouces",
                                   ofType: "bundle")
        
        guard let bundle = path.flatMap({ Bundle(path: $0) }) else {
            fatalError("MyPrecompiledFrameworkResouces could not be loaded")
        }
        
        return bundle
  }
```

I wonder if there's a better way to deal with this overall

**Comment by RomainBoulay** [comment]
Thanks a lot @kwridan for taking the time to reply with such a level of details, it's really appreciated 🙏

I agree with you that our approach does not probably represent what other users may experience. We've followed this one to make sure our build times don't rise too much, since we do need to keep adding more and more dependencies over time.

As things stand, we don't really need to access `FrameworkA` resources, we just have a unit test to make sure there is no runtime crash when `FrameworkA` accesses its own resources. So the simlink workaround is probably the best, cause the lightest.

Your idea to create a bundle with `FrameworkA` resources is very nice, but I'm afraid that's too much build times added, and development effort given we just have 1 unit test that require `File1.js`.

To sum up, thanks again @kwridan! you helped to clarify our issue and that was very insightful.
